FrEee/Scripts/AI/Vanguard/Plan.csx

```diff
@@ -0,0 +1,278 @@
+        CurrentPlans.ShipBuildPlans.RemoveAll(x => x.IsComplete); 
+    }
+
+    
+}
+
+/// <summary>
+/// The method of storing plans between turns for the AI. 
+/// </summary>
+[Serializable]
+public class PlanList
+{
+    public List<ColonizationPlan> ColonizationPlans;
+
+    public List<ShipBuildPlan> ShipBuildPlans; 
+
+    [DoNotSerialize]
+    public IEnumerable<Plan> AllPlans { 
+        get
+        {
+            var plans = new List<Plan>();
+            plans.AddRange(ColonizationPlans);
+            plans.AddRange(ShipBuildPlans);
+            return plans; 
+        } 
+    }
+
+
+    public PlanList()
+    {
+        ColonizationPlans = new List<ColonizationPlan>();
+        ShipBuildPlans = new List<ShipBuildPlan>(); 
+    }
+
+    public void AssignAllIds()
+    {
+        foreach (var p in ColonizationPlans)
+            if (p.PlanId == null)
+                p.AssignId();
+        foreach (var p in ShipBuildPlans)
+            if (p.PlanId == null)
+                p.AssignId();
+    }
+
+    /// <summary>
+    /// gets the chosen plan, or null. 
+    /// </summary>
+    /// <param name="planId"></param>
+    /// <returns></returns>
+    public Plan GetPlan(string planId)
+    {
+        return AllPlans.FirstOrDefault(x => x.PlanId == planId); 
+    }
+}
+
+
+[Serializable]
+public abstract class Plan
+{
+    public string PlanId; 
+
+
+    [DoNotSerialize]
+    public virtual string TypeShorthand { get; }
+
+    public bool IsComplete;
+
+    /// <summary>
+    /// The priority of the plan. Higher is better.  
+    /// </summary>
+    public int Priority; 
+
+
+    public void AssignId()
+    {
+        PlanId = $"{TypeShorthand}-{DateTime.UtcNow.Ticks}-{PlanManager.CurrentPlans.AllPlans.Count()}";
+    }
+}
+
+
+[Serializable]
+public class ColonizationPlan : Plan
+{
+    [DoNotSerialize]
+    public override string TypeShorthand => "C";
+
+    /// <summary>
+    /// The planet intended to colonize. 
+    /// </summary>
+    [DoNotSerialize]
+    public Planet Planet { get {
+            if (PlanetId == 0)
+                return null; return GalaxyReference<Planet>.GetGalaxyReference(PlanetId);  } set { PlanetId = value.ID; } }
+
+    /// <summary>
+    /// The planet that will build the colony ship to colonize this world. 
+    /// </summary>
+    [DoNotSerialize]
+    public Planet SourcePlanet { get
+        {
+            if (SourcePlanetId == 0)
+                return null; return GalaxyReference<Planet>.GetGalaxyReference(SourcePlanetId); } set { SourcePlanetId = value.ID; } }
+
+    [DoNotSerialize]
+    public ShipBuildPlan ShipBuildPlan
+    {
+        get
+        {
+            if (ShipBuildPlanId == null)
+                return null;
+
+            return PlanManager.CurrentPlans.ShipBuildPlans.FirstOrDefault(x => x.PlanId == ShipBuildPlanId); 
+        }
+    }
+
+
+    /// <summary>
+    /// The Id of the planet. 
+    /// </summary>
+    public long PlanetId;
```

Comment:
All these IDs - you don't want to use the GalaxyReference class, or create a new subclass of IReference?

Reply:
I started out not using them without realizing the existance of the GalaxyReference class. I suppose I could add a new subclass, though it only needs the long ID, so wouldn't it increase the amoutn of data being written to the AI memory that way?

Reply:
A reference is just a wrapper around an ID that has methods to look up the actual object and such. I think it would store a bit more data than just the ID, if only some metadata on the Reference type itself. Maybe we could save space in the savegame by getting rid of References and instead adding the lookup as an extension method on the `long` type? Do whatever works best for you, I guess...